presto-main/src/main/java/io/prestosql/sql/planner/TranslationMap.java

```diff
@@ -47,181 +54,162 @@
-
-        fieldSymbols = new Symbol[rewriteBase.getFieldMappings().size()];
-    }
-
-    public RelationPlan getRelationPlan()
-    {
-        return rewriteBase;
+        this(outerContext, scope, analysis, lambdaArguments, fieldSymbols.toArray(new Symbol[0]).clone(), ImmutableMap.of());
     }
 
-    public Analysis getAnalysis()
+    public TranslationMap(Optional<TranslationMap> outerContext, Scope scope, Analysis analysis, Map<NodeRef<LambdaArgumentDeclaration>, Symbol> lambdaArguments, List<Symbol> fieldSymbols, Map<ScopeAware<Expression>, Symbol> astToSymbols)
     {
-        return analysis;
+        this(outerContext, scope, analysis, lambdaArguments, fieldSymbols.toArray(new Symbol[0]), astToSymbols);
     }
 
-    public Map<NodeRef<LambdaArgumentDeclaration>, Symbol> getLambdaDeclarationToSymbolMap()
+    public TranslationMap(Optional<TranslationMap> outerContext, Scope scope, Analysis analysis, Map<NodeRef<LambdaArgumentDeclaration>, Symbol> lambdaArguments, Symbol[] fieldSymbols, Map<ScopeAware<Expression>, Symbol> astToSymbols)
     {
-        return lambdaDeclarationToSymbolMap;
-    }
+        this.outerContext = requireNonNull(outerContext, "outerContext is null");
+        this.scope = requireNonNull(scope, "scope is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.lambdaDeclarationToSymbolMap = requireNonNull(lambdaArguments, "lambdaArguments is null");
 
-    public void setFieldMappings(List<Symbol> symbols)
-    {
-        checkArgument(symbols.size() == fieldSymbols.length, "size of symbols list (%s) doesn't match number of expected fields (%s)", symbols.size(), fieldSymbols.length);
+        requireNonNull(fieldSymbols, "fieldSymbols is null");
+        this.fieldSymbols = fieldSymbols.clone();
 
-        for (int i = 0; i < symbols.size(); i++) {
-            this.fieldSymbols[i] = symbols.get(i);
-        }
-    }
+        requireNonNull(astToSymbols, "astToSymbols is null");
+        this.astToSymbols = ImmutableMap.copyOf(astToSymbols);
 
-    public void copyMappingsFrom(TranslationMap other)
-    {
-        checkArgument(other.fieldSymbols.length == fieldSymbols.length,
-                "number of fields in other (%s) doesn't match number of expected fields (%s)",
-                other.fieldSymbols.length,
+        checkArgument(scope.getLocalScopeFieldCount() == fieldSymbols.length,
+                "scope: %s, fields mappings: %s",
+                scope.getRelationType().getAllFieldCount(),
                 fieldSymbols.length);
 
-        expressionToSymbols.putAll(other.expressionToSymbols);
-        expressionToExpressions.putAll(other.expressionToExpressions);
-        System.arraycopy(other.fieldSymbols, 0, fieldSymbols, 0, other.fieldSymbols.length);
+        astToSymbols.keySet().stream()
+            .map(ScopeAware::getNode)
+            .forEach(TranslationMap::verifyAstExpression);
```

Comment:
I think this verify deserves a comment, either here or at the declaration of `astToSymbols` field.
It should say that TranslationMap only holds mappings for purely-ast expressions.
And some explanation what those are, e.g. what are the ast-land references to columns as opposed to IR-land `SymbolReference`.
Using TranslationMap with non-ast expressions fails, either when trying to add mappings or at call to `canTranslate()` or `rewrite()`.

Reply:
Ok, I'll add a comment in the class Javadocs